// found/esm/createElements.js
import _mapInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/map";
import React from 'react';
import warning from 'warning';
import { isResolved } from './ResolverUtils';
export default function createElements(routeMatches, Components, matchData) {
  return _mapInstanceProperty(routeMatches).call(routeMatches, function (match, i) {
    var router = match.router,
        route = match.route;
    var Component = Components[i];
    var data = matchData[i];
    var isComponentResolved = isResolved(Component);
    var areDataResolved = isResolved(data);

    if (route.render) {
      // Perhaps undefined here would be more correct for "not ready", but
      // Relay uses null in RelayReadyStateRenderer, so let's follow that
      // convention.
      return route.render({
        match: match,
        Component: isComponentResolved ? Component : null,
        props: areDataResolved ? {
          match: match,
          router: router,
          data: data
        } : null,
        data: areDataResolved ? data : null
      });
    }

    if (!isComponentResolved || !areDataResolved) {
      // Can't render.
      return undefined;
    }

    if (!Component) {
      // Note this check would be wrong on potentially unresolved data.
      process.env.NODE_ENV !== "production" ? warning(data === undefined, 'Route %s with data has no render method or component.', i) : void 0; // Nothing to render.

      return null;
    }

    return /*#__PURE__*/React.createElement(Component, {
      match: match,
      router: router,
      data: data
    });
  });
}
