// found/esm/resolveRenderArgs.js
import _bindInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/bind";
import _getIterator from "@babel/runtime-corejs3/core-js/get-iterator";
import _Array$isArray from "@babel/runtime-corejs3/core-js-stable/array/is-array";
import _getIteratorMethod from "@babel/runtime-corejs3/core-js/get-iterator-method";
import _Symbol from "@babel/runtime-corejs3/core-js-stable/symbol";
import _Array$from from "@babel/runtime-corejs3/core-js-stable/array/from";
import _sliceInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/slice";
import _regeneratorRuntime from "@babel/runtime-corejs3/regenerator";
import _concatInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/concat";
import _extends from "@babel/runtime-corejs3/helpers/esm/extends";
import _Object$entries from "@babel/runtime-corejs3/core-js-stable/object/entries";
import _forEachInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/for-each";
import _awaitAsyncGenerator from "@babel/runtime-corejs3/helpers/esm/awaitAsyncGenerator";
import _wrapAsyncGenerator from "@babel/runtime-corejs3/helpers/esm/wrapAsyncGenerator";
import _asyncIterator from "@babel/runtime-corejs3/helpers/esm/asyncIterator";

function _createForOfIteratorHelperLoose(o, allowArrayLike) { var _context5; var it; if (typeof _Symbol === "undefined" || _getIteratorMethod(o) == null) { if (_Array$isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; return function () { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } it = _getIterator(o); return _bindInstanceProperty(_context5 = it.next).call(_context5, it); }

function _unsupportedIterableToArray(o, minLen) { var _context4; if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = _sliceInstanceProperty(_context4 = Object.prototype.toString.call(o)).call(_context4, 8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return _Array$from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

import HttpError from './HttpError';

function foldElements(elementsRaw, routeIndices) {
  var elements = [];

  for (var _iterator2 = _createForOfIteratorHelperLoose(routeIndices), _step2; !(_step2 = _iterator2()).done;) {
    var routeIndex = _step2.value;

    if (typeof routeIndex === 'object') {
      (function () {
        var _context;

        // Reshape the next elements in the elements array to match the nested
        // tree structure corresponding to the route groups.
        var groupElements = {};

        _forEachInstanceProperty(_context = _Object$entries(routeIndex)).call(_context, function (_ref) {
          var groupName = _ref[0],
              groupRouteIndices = _ref[1];
          groupElements[groupName] = foldElements(elementsRaw, groupRouteIndices);
        });

        elements.push(groupElements);
      })();
    } else {
      // We intentionally modify elementsRaw, to make it easier to recursively
      // handle groups.
      elements.push(elementsRaw.shift());
    }
  }

  return elements;
}

export default function resolveRenderArgs(_x, _x2) {
  return _resolveRenderArgs.apply(this, arguments);
}

function _resolveRenderArgs() {
  _resolveRenderArgs = _wrapAsyncGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee(router, _ref2) {
    var match, matchContext, resolver, routes, augmentedMatch, _iteratorNormalCompletion, _didIteratorError, _iteratorError, _iterator, _step, _value, _context2, elements;

    return _regeneratorRuntime.wrap(function _callee$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            match = _ref2.match, matchContext = _ref2.matchContext, resolver = _ref2.resolver;
            routes = router.matcher.getRoutes(match);
            augmentedMatch = _extends({}, match, {
              routes: routes,
              router: router,
              // Convenience access for route components.
              context: matchContext
            });

            if (routes) {
              _context3.next = 7;
              break;
            }

            _context3.next = 6;
            return _extends({}, augmentedMatch, {
              error: new HttpError(404)
            });

          case 6:
            return _context3.abrupt("return");

          case 7:
            _context3.prev = 7;
            _iteratorNormalCompletion = true;
            _didIteratorError = false;
            _context3.prev = 10;
            _iterator = _asyncIterator(resolver.resolveElements(augmentedMatch));

          case 12:
            _context3.next = 14;
            return _awaitAsyncGenerator(_iterator.next());

          case 14:
            _step = _context3.sent;
            _iteratorNormalCompletion = _step.done;
            _context3.next = 18;
            return _awaitAsyncGenerator(_step.value);

          case 18:
            _value = _context3.sent;

            if (_iteratorNormalCompletion) {
              _context3.next = 26;
              break;
            }

            elements = _value;
            _context3.next = 23;
            return _extends({}, augmentedMatch, {
              elements: elements && foldElements(_concatInstanceProperty(_context2 = []).call(_context2, elements), match.routeIndices)
            });

          case 23:
            _iteratorNormalCompletion = true;
            _context3.next = 12;
            break;

          case 26:
            _context3.next = 32;
            break;

          case 28:
            _context3.prev = 28;
            _context3.t0 = _context3["catch"](10);
            _didIteratorError = true;
            _iteratorError = _context3.t0;

          case 32:
            _context3.prev = 32;
            _context3.prev = 33;

            if (!(!_iteratorNormalCompletion && _iterator["return"] != null)) {
              _context3.next = 37;
              break;
            }

            _context3.next = 37;
            return _awaitAsyncGenerator(_iterator["return"]());

          case 37:
            _context3.prev = 37;

            if (!_didIteratorError) {
              _context3.next = 40;
              break;
            }

            throw _iteratorError;

          case 40:
            return _context3.finish(37);

          case 41:
            return _context3.finish(32);

          case 42:
            _context3.next = 51;
            break;

          case 44:
            _context3.prev = 44;
            _context3.t1 = _context3["catch"](7);

            if (!_context3.t1.isFoundHttpError) {
              _context3.next = 50;
              break;
            }

            _context3.next = 49;
            return _extends({}, augmentedMatch, {
              error: _context3.t1
            });

          case 49:
            return _context3.abrupt("return");

          case 50:
            throw _context3.t1;

          case 51:
          case "end":
            return _context3.stop();
        }
      }
    }, _callee, null, [[7, 44], [10, 28, 32, 42], [33,, 37, 41]]);
  }));
  return _resolveRenderArgs.apply(this, arguments);
}
