// farce/esm/BrowserProtocol.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import invariant from 'invariant';
import createPath from './createPath';

var BrowserProtocol = /*#__PURE__*/function () {
  function BrowserProtocol() {
    this._keyPrefix = Math.random().toString(36).slice(2, 8);
    this._keyIndex = 0;
    this._index = null;
  }

  var _proto = BrowserProtocol.prototype;

  _proto.init = function init() {
    var _window$location = window.location,
        pathname = _window$location.pathname,
        search = _window$location.search,
        hash = _window$location.hash;

    var _ref = window.history.state || {},
        key = _ref.key,
        _ref$index = _ref.index,
        index = _ref$index === void 0 ? 0 : _ref$index,
        state = _ref.state;

    var delta = this._index != null ? index - this._index : 0;
    this._index = index;
    return {
      action: 'POP',
      pathname: pathname,
      search: search,
      hash: hash,
      key: key,
      index: index,
      delta: delta,
      state: state
    };
  };

  _proto.subscribe = function subscribe(listener) {
    var _this = this;

    var onPopState = function onPopState() {
      listener(_this.init());
    }; // TODO: On most versions of IE, we need a hashChange listener for hash-
    //  only changes.


    window.addEventListener('popstate', onPopState);
    return function () {
      return window.removeEventListener('popstate', onPopState);
    };
  };

  _proto.navigate = function navigate(location) {
    var action = location.action,
        state = location.state;
    var push = action === 'PUSH';
    !(push || action === 'REPLACE') ? process.env.NODE_ENV !== "production" ? invariant(false, "Unrecognized browser protocol action: %s.", action) : invariant(false) : void 0;
    var delta = push ? 1 : 0;

    var extraState = this._createExtraState(delta);

    var browserState = _extends({
      state: state
    }, extraState);

    var path = createPath(location);

    if (push) {
      window.history.pushState(browserState, null, path);
    } else {
      window.history.replaceState(browserState, null, path);
    }

    return _extends({}, location, extraState, {
      delta: delta
    });
  };

  _proto.go = function go(delta) {
    window.history.go(delta);
  };

  _proto.createHref = function createHref(location) {
    return createPath(location);
  };

  _proto._createExtraState = function _createExtraState(delta) {
    var keyIndex = this._keyIndex++;
    this._index += delta;
    return {
      key: this._keyPrefix + ":" + keyIndex.toString(36),
      index: this._index
    };
  };

  return BrowserProtocol;
}();

export { BrowserProtocol as default };
