// found/cjs/createBaseRouter.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = createBaseRouter;

var _regenerator = _interopRequireDefault(require("@babel/runtime-corejs3/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/asyncToGenerator"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/inheritsLoose"));

var _asyncIterator2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/asyncIterator"));

var _mapContextToProps = _interopRequireDefault(require("@restart/context/mapContextToProps"));

var _isEqual = _interopRequireDefault(require("lodash/isEqual"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _reactRedux = require("react-redux");

var _reactStaticContainer = _interopRequireDefault(require("react-static-container"));

var _warning = _interopRequireDefault(require("warning"));

var _RouterContext = _interopRequireDefault(require("./RouterContext"));

var _createRender = _interopRequireDefault(require("./createRender"));

var _createStoreRouterObject = _interopRequireDefault(require("./createStoreRouterObject"));

var _resolveRenderArgs = _interopRequireDefault(require("./resolveRenderArgs"));

function createBaseRouter(_ref) {
  var renderPending = _ref.renderPending,
      renderReady = _ref.renderReady,
      renderError = _ref.renderError,
      _ref$render = _ref.render,
      render = _ref$render === void 0 ? (0, _createRender["default"])({
    renderPending: renderPending,
    renderReady: renderReady,
    renderError: renderError
  }) : _ref$render;
  var propTypes = {
    store: _propTypes["default"].object.isRequired,
    match: _propTypes["default"].object.isRequired,
    resolvedMatch: _propTypes["default"].object.isRequired,
    matchContext: _propTypes["default"].any,
    resolver: _propTypes["default"].shape({
      resolveElements: _propTypes["default"].func.isRequired
    }).isRequired,
    onResolveMatch: _propTypes["default"].func.isRequired,
    initialRenderArgs: _propTypes["default"].object
  };

  var BaseRouter = /*#__PURE__*/function (_React$Component) {
    (0, _inheritsLoose2["default"])(BaseRouter, _React$Component);

    function BaseRouter(props) {
      var _this;

      _this = _React$Component.call(this, props) || this;
      var store = props.store,
          match = props.match,
          matchContext = props.matchContext,
          resolver = props.resolver,
          initialRenderArgs = props.initialRenderArgs;
      _this.router = (0, _createStoreRouterObject["default"])(store);
      _this.state = {
        isInitialRender: true,
        match: match,
        matchContext: matchContext,
        resolver: resolver,
        iteration: 0,
        routerContext: {
          router: _this.router,
          match: initialRenderArgs || null
        },
        element: initialRenderArgs ? render(initialRenderArgs) : null
      };
      _this.mounted = true;
      _this.lastIteration = 0;
      _this.pendingResolvedMatch = false;
      return _this;
    } // We use componentDidMount and componentDidUpdate to resolve the match if
    //  needed because element resolution is asynchronous anyway, and this lets
    //  us not worry about setState not being available in the constructor, or
    //  about having to pass around nextProps.


    var _proto = BaseRouter.prototype;

    _proto.componentDidMount = function componentDidMount() {
      if (!this.props.initialRenderArgs) {
        this.resolveMatch();
      }

      if (process.env.NODE_ENV !== "production" && typeof window !== 'undefined') {
        /* eslint-env browser */

        /* eslint-disable no-underscore-dangle */
        if (window.__FOUND_HOT_RELOAD__) {
          process.env.NODE_ENV !== "production" ? (0, _warning["default"])(!window.__FOUND_REPLACE_ROUTE_CONFIG__, 'Replacing existing hot reloading hook. Do not render more than one router instance when using hot reloading.') : void 0;
          window.__FOUND_REPLACE_ROUTE_CONFIG__ = this.router.replaceRouteConfig;
        }
        /* eslint-enable no-underscore-dangle */

        /* eslint-env browser: false */

      }
    };

    BaseRouter.getDerivedStateFromProps = function getDerivedStateFromProps(_ref2, state) {
      var match = _ref2.match,
          resolver = _ref2.resolver,
          matchContext = _ref2.matchContext;

      if (state.isInitialRender) {
        return {
          isInitialRender: false
        };
      }

      if (match !== state.match || resolver !== state.resolver || !(0, _isEqual["default"])(matchContext, state.matchContext)) {
        return {
          match: match,
          resolver: resolver,
          matchContext: matchContext,
          iteration: state.iteration + 1
        };
      }

      return null;
    };

    _proto.componentDidUpdate = function componentDidUpdate() {
      if (this.state.iteration > this.lastIteration) {
        this.lastIteration = this.state.iteration;
        this.resolveMatch();
      }
    };

    _proto.componentWillUnmount = function componentWillUnmount() {
      this.mounted = false;

      if (process.env.NODE_ENV !== "production" && typeof window !== 'undefined') {
        /* eslint-env browser */

        /* eslint-disable no-underscore-dangle */
        if (window.__FOUND_HOT_RELOAD__) {
          delete window.__FOUND_REPLACE_ROUTE_CONFIG__;
        }
        /* eslint-enable no-underscore-dangle */

        /* eslint-env browser: false */

      }
    };

    _proto.resolveMatch = /*#__PURE__*/function () {
      var _resolveMatch = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
        var pendingIteration, pendingMatch, _iteratorNormalCompletion, _didIteratorError, _iteratorError, _iterator, _step, _value, renderArgs;

        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                pendingIteration = this.lastIteration;
                pendingMatch = this.props.match;
                _context.prev = 2;
                _iteratorNormalCompletion = true;
                _didIteratorError = false;
                _context.prev = 5;
                _iterator = (0, _asyncIterator2["default"])((0, _resolveRenderArgs["default"])(this.router, this.props));

              case 7:
                _context.next = 9;
                return _iterator.next();

              case 9:
                _step = _context.sent;
                _iteratorNormalCompletion = _step.done;
                _context.next = 13;
                return _step.value;

              case 13:
                _value = _context.sent;

                if (_iteratorNormalCompletion) {
                  _context.next = 24;
                  break;
                }

                renderArgs = _value;

                if (!(!this.mounted || this.lastIteration !== pendingIteration)) {
                  _context.next = 18;
                  break;
                }

                return _context.abrupt("return");

              case 18:
                // If we're about to mark the match resolved, delay the rerender
                //  until we do so.
                this.pendingResolvedMatch = !!((renderArgs.elements || renderArgs.error) && this.props.resolvedMatch !== pendingMatch);
                this.setState({
                  routerContext: {
                    router: this.router,
                    match: renderArgs
                  },
                  element: render(renderArgs)
                });

                if (this.pendingResolvedMatch) {
                  // If this is a new match, update the store, so we can rerender at
                  //  the same time as all of the links and other components
                  //  connected to the router state.
                  this.pendingResolvedMatch = false;
                  this.props.onResolveMatch(pendingMatch);
                }

              case 21:
                _iteratorNormalCompletion = true;
                _context.next = 7;
                break;

              case 24:
                _context.next = 30;
                break;

              case 26:
                _context.prev = 26;
                _context.t0 = _context["catch"](5);
                _didIteratorError = true;
                _iteratorError = _context.t0;

              case 30:
                _context.prev = 30;
                _context.prev = 31;

                if (!(!_iteratorNormalCompletion && _iterator["return"] != null)) {
                  _context.next = 35;
                  break;
                }

                _context.next = 35;
                return _iterator["return"]();

              case 35:
                _context.prev = 35;

                if (!_didIteratorError) {
                  _context.next = 38;
                  break;
                }

                throw _iteratorError;

              case 38:
                return _context.finish(35);

              case 39:
                return _context.finish(30);

              case 40:
                _context.next = 50;
                break;

              case 42:
                _context.prev = 42;
                _context.t1 = _context["catch"](2);

                if (!(!this.mounted || this.lastIteration !== pendingIteration)) {
                  _context.next = 46;
                  break;
                }

                return _context.abrupt("return");

              case 46:
                if (!_context.t1.isFoundRedirectException) {
                  _context.next = 49;
                  break;
                }

                this.router.replace(_context.t1.location);
                return _context.abrupt("return");

              case 49:
                throw _context.t1;

              case 50:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this, [[2, 42], [5, 26, 30, 40], [31,, 35, 39]]);
      }));

      function resolveMatch() {
        return _resolveMatch.apply(this, arguments);
      }

      return resolveMatch;
    }();

    _proto.render = function render() {
      var _this$state = this.state,
          iteration = _this$state.iteration,
          routerContext = _this$state.routerContext,
          element = _this$state.element; // Don't rerender synchronously if we have another rerender coming.

      return /*#__PURE__*/_react["default"].createElement(_reactStaticContainer["default"], {
        shouldUpdate: this.lastIteration === iteration && !this.pendingResolvedMatch
      }, /*#__PURE__*/_react["default"].createElement(_RouterContext["default"].Provider, {
        value: routerContext
      }, element));
    };

    return BaseRouter;
  }(_react["default"].Component);

  BaseRouter.propTypes = propTypes; // FIXME: For some reason, using contextType doesn't work here.

  return (0, _mapContextToProps["default"])({
    consumers: _reactRedux.ReactReduxContext,
    mapToProps: function mapToProps(_ref3) {
      var store = _ref3.store;
      return {
        store: store
      };
    },
    displayName: 'withStore(BaseRouter)'
  }, BaseRouter);
}

module.exports = exports.default;
