// found/cjs/PropTypes.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

exports.__esModule = true;
exports.routerShape = exports.matcherShape = exports.matchShape = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

// The shape might be different with a custom matcher or history enhancer, but
// the default matcher assumes and provides this shape. As such, this validator
// is purely for user convenience and should not be used internally.
var matchShape = _propTypes["default"].shape({
  location: _propTypes["default"].shape({
    pathname: _propTypes["default"].string.isRequired,
    query: _propTypes["default"].object.isRequired
  }).isRequired,
  params: _propTypes["default"].object.isRequired
}); // User code generally shouldn't need this, but it doesn't hurt to export here,
// since we use it for routerShape below.


exports.matchShape = matchShape;

var matcherShape = _propTypes["default"].shape({
  match: _propTypes["default"].func.isRequired,
  getRoutes: _propTypes["default"].func.isRequired,
  isActive: _propTypes["default"].func.isRequired,
  format: _propTypes["default"].func.isRequired
});

exports.matcherShape = matcherShape;

var routerShape = _propTypes["default"].shape({
  push: _propTypes["default"].func.isRequired,
  replace: _propTypes["default"].func.isRequired,
  go: _propTypes["default"].func.isRequired,
  createHref: _propTypes["default"].func.isRequired,
  createLocation: _propTypes["default"].func.isRequired,
  isActive: _propTypes["default"].func.isRequired,
  matcher: matcherShape.isRequired,
  addNavigationListener: _propTypes["default"].func.isRequired
});

exports.routerShape = routerShape;
