// found/cjs/resolver.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _concat = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/concat"));

var _regenerator = _interopRequireDefault(require("@babel/runtime-corejs3/regenerator"));

var _every = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/every"));

var _map = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/map"));

var _promise = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/promise"));

var _some = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/some"));

var _awaitAsyncGenerator2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/awaitAsyncGenerator"));

var _wrapAsyncGenerator2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/wrapAsyncGenerator"));

var _isPromise = _interopRequireDefault(require("is-promise"));

var _ResolverUtils = require("./ResolverUtils");

var _createElements = _interopRequireDefault(require("./createElements"));

function getRouteGetData(route) {
  return route.getData;
}

function getRouteData(route) {
  return route.data;
}

var _default = {
  resolveElements: function resolveElements(match) {
    var _this = this;

    return (0, _wrapAsyncGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
      var routeMatches, Components, data, earlyComponents, earlyData, fetchedComponents, fetchedData, pendingElements;
      return _regenerator["default"].wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              routeMatches = (0, _ResolverUtils.getRouteMatches)(match);
              Components = (0, _ResolverUtils.getComponents)(routeMatches);
              data = _this.getData(match, routeMatches);

              if (!(0, _some["default"])(Components).call(Components, _isPromise["default"])) {
                _context.next = 9;
                break;
              }

              _context.next = 6;
              return (0, _awaitAsyncGenerator2["default"])(_promise["default"].all((0, _map["default"])(Components).call(Components, _ResolverUtils.checkResolved)));

            case 6:
              _context.t0 = _context.sent;
              _context.next = 10;
              break;

            case 9:
              _context.t0 = Components;

            case 10:
              earlyComponents = _context.t0;

              if (!(0, _some["default"])(data).call(data, _isPromise["default"])) {
                _context.next = 17;
                break;
              }

              _context.next = 14;
              return (0, _awaitAsyncGenerator2["default"])(_promise["default"].all((0, _map["default"])(data).call(data, _ResolverUtils.checkResolved)));

            case 14:
              _context.t1 = _context.sent;
              _context.next = 18;
              break;

            case 17:
              _context.t1 = data;

            case 18:
              earlyData = _context.t1;

              if (!(!(0, _every["default"])(earlyComponents).call(earlyComponents, _ResolverUtils.isResolved) || !(0, _every["default"])(earlyData).call(earlyData, _ResolverUtils.isResolved))) {
                _context.next = 31;
                break;
              }

              pendingElements = (0, _createElements["default"])(routeMatches, earlyComponents, earlyData);
              _context.next = 23;
              return (0, _every["default"])(pendingElements).call(pendingElements, function (element) {
                return element !== undefined;
              }) ? pendingElements : undefined;

            case 23:
              _context.next = 25;
              return (0, _awaitAsyncGenerator2["default"])(_promise["default"].all(Components));

            case 25:
              fetchedComponents = _context.sent;
              _context.next = 28;
              return (0, _awaitAsyncGenerator2["default"])(_promise["default"].all(data));

            case 28:
              fetchedData = _context.sent;
              _context.next = 33;
              break;

            case 31:
              fetchedComponents = earlyComponents;
              fetchedData = earlyData;

            case 33:
              _context.next = 35;
              return (0, _createElements["default"])(routeMatches, fetchedComponents, fetchedData);

            case 35:
            case "end":
              return _context.stop();
          }
        }
      }, _callee);
    }))();
  },

  /**
   * Generate route data according to their getters, respecting the order of
   * promises per the `defer` flag on routes.
   */
  getData: function getData(match, routeMatches) {
    var _context2;

    return (0, _map["default"])(_context2 = (0, _ResolverUtils.accumulateRouteValues)(routeMatches, match.routeIndices, function (_ref, routeMatch) {
      var _context3;

      var ancestorRouteData = _ref.ancestorRouteData,
          prevParentPromise = _ref.prevParentPromise;
      // For a deferred route, the parent promise is the previous promise.
      // Otherwise, it's the previous parent promise.
      var parentPromise = routeMatch.route.defer ? _promise["default"].all(ancestorRouteData) : prevParentPromise; // If there is a parent promise, execute after it resolves.

      var routeData = parentPromise ? parentPromise.then(function () {
        return (0, _ResolverUtils.getRouteValue)(routeMatch, getRouteGetData, getRouteData);
      }) : (0, _ResolverUtils.getRouteValue)(routeMatch, getRouteGetData, getRouteData);
      return {
        routeData: routeData,
        ancestorRouteData: (0, _concat["default"])(_context3 = []).call(_context3, ancestorRouteData, [routeData]),
        prevParentPromise: parentPromise
      };
    }, {
      routeData: null,
      ancestorRouteData: [],
      prevParentPromise: null
    })).call(_context2, function (_ref2) {
      var routeData = _ref2.routeData;
      return routeData;
    });
  }
};
exports["default"] = _default;
module.exports = exports.default;
